packages/hardhat-chai-matchers/src/reverted/revertedWithCustomError.ts

```diff
@@ -0,0 +1,230 @@
+        );
+      }
+
+      const iface: any = contract.interface;
+
+      const expectedCustomError = findCustomErrorByName(
+        iface,
+        expectedCustomErrorName
+      );
+
+      // check that interface contains the given custom error
+      if (expectedCustomError === undefined) {
+        throw new AssertionError(
+          `The given contract doesn't have a custom error named ${expectedCustomErrorName}`
+        );
+      }
+
+      const onSuccess = () => {
+        this.assert(
+          false,
+          `Expected transaction to be reverted with custom error '${expectedCustomErrorName}', but it didn't revert`
+        );
+      };
+
+      const onError = (error: any) => {
+        const returnData = getReturnDataFromError(error);
+        const decodedReturnData = decodeReturnData(returnData);
+
+        if (decodedReturnData.kind === "Empty") {
+          this.assert(
+            false,
+            `Expected transaction to be reverted with custom error '${expectedCustomErrorName}', but it reverted without a reason string`
+          );
+        } else if (decodedReturnData.kind === "Error") {
+          this.assert(
+            false,
+            `Expected transaction to be reverted with custom error '${expectedCustomErrorName}', but it reverted with reason '${decodedReturnData.reason}'`
+          );
+        } else if (decodedReturnData.kind === "Panic") {
+          this.assert(
+            false,
+            `Expected transaction to be reverted with custom error '${expectedCustomErrorName}', but it reverted with panic code ${decodedReturnData.code.toHexString()} (${
+              decodedReturnData.description
+            })`
+          );
+        } else if (decodedReturnData.kind === "Custom") {
+          if (decodedReturnData.id === expectedCustomError.id) {
+            // add flag with the data needed for .withArgs
+            const customErrorAssertionData: CustomErrorAssertionData = {
+              contractInterface: iface,
+              customError: expectedCustomError,
+              returnData,
+            };
+            utils.flag(
+              this,
+              CUSTOM_ERROR_ASSERTION_DATA,
+              customErrorAssertionData
+            );
+
+            this.assert(
+              true,
+              null,
+              `Expected transaction NOT to be reverted with custom error '${expectedCustomErrorName}', but it did`
+            );
+          } else {
+            // try to decode the actual custom error
+            // this will only work when the error comes from the given contract
+            const actualCustomError = findCustomErrorById(
+              iface,
+              decodedReturnData.id
+            );
+
+            if (actualCustomError === undefined) {
+              this.assert(
+                false,
+                `Expected transaction to be reverted with custom error '${expectedCustomErrorName}', but it reverted with a different custom error`
```

Comment:
Perhaps be more clear here that the "different custom error" must be from a different contract, and perhaps also mention what was tried (eg "looked for it in this contract but it didn't exist, therefore it must be from a different contract").

Reply:
What do you think of:

`Expected transaction to be reverted with custom error '${expectedCustomErrorName}', but it reverted with a different custom error. This usually happens because the transaction reverted with a custom error in another contract.`